db/commitlog/commitlog.cc

```diff
@@ -619,6 +619,47 @@ std::enable_if_t<std::is_fundamental<T>::value, T> read(Input& in) {
     return net::ntoh(in.template read<T>());
 }
 
+detail::sector_split_iterator::sector_split_iterator(const sector_split_iterator&) noexcept = default;
+
+detail::sector_split_iterator::sector_split_iterator()
+    : _ptr(nullptr)
+    , _size(0)
+    , _sector_size(0)
+{}
+
+detail::sector_split_iterator::sector_split_iterator(base_iterator i, base_iterator e, size_t sector_size)
+    : _iter(i)
+    , _end(e)
+    , _ptr(i != e ? const_cast<char*>(i->get()) : nullptr)
+    , _size(i != e ? sector_size - sector_overhead_size : 0)
+    , _sector_size(sector_size)
+{}
+
+detail::sector_split_iterator& detail::sector_split_iterator::operator++() {
+    assert(_iter != _end);
+    _ptr += _sector_size;
+    // check if we have more pages in this temp-buffer (in out case they are always aligned + sized in page units)
+    auto rem = _iter->size() - std::distance(_iter->get(), const_cast<const char*>(_ptr));
```

Comment:
That is more C thinking. The above is (imho) more clear, because we are subtracting the distance between pointers from a size, i.e. different types. std::distance handles that "cast" part for us, i.e. no type mixing. (Except for the ugly const-cast).

Reply:
There is std::as_const()